bundles/org.eclipse.cdt.lsp.clangd/src/org/eclipse/cdt/lsp/internal/clangd/ClangdConfigFileChecker.java

```diff
@@ -0,0 +1,115 @@
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.error.MarkedYAMLException;
+
+/**
+ * Checks the <code>.clangd</code> file for syntax errors and notifies the user via error markers in the file and Problems view.
+ */
+public class ClangdConfigFileChecker {
+	public static final String CLANGD_MARKER = ClangdPlugin.PLUGIN_ID + ".config.marker"; //$NON-NLS-1$
+
+	/**
+	 * Checks if the .clangd file contains valid yaml syntax. Adds error marker to the file if not.
+	 * @param configFile
+	 */
+	public void checkConfigFile(IFile configFile) {
+		Yaml yaml = new Yaml();
+		try (var inputStream = configFile.getContents()) {
+			try {
+				removeMarkerFromClangdConfig(configFile);
+				//throws ScannerException and ParserException:
+				yaml.load(inputStream);
+			} catch (Exception exception) {
+				if (exception instanceof MarkedYAMLException yamlException) {
+					addMarkerToClangdConfig(configFile, yamlException);
+				} else {
+					//log unexpected exception:
+					Platform.getLog(getClass())
+							.error("Expected MarkedYAMLException, but was: " + exception.getMessage(), exception); //$NON-NLS-1$
+				}
+			}
+		} catch (IOException | CoreException e) {
+			Platform.getLog(getClass()).error(e.getMessage(), e);
+		}
+	}
+
+	private void addMarkerToClangdConfig(IFile configFile, MarkedYAMLException yamlException) {
+		try {
+			var configMarker = parseYamlException(yamlException);
+			var marker = configFile.createMarker(CLANGD_MARKER);
+			marker.setAttribute(IMarker.MESSAGE, configMarker.message);
+			marker.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
+			marker.setAttribute(IMarker.LINE_NUMBER, configMarker.line);
+			marker.setAttribute(IMarker.CHAR_START, configMarker.charStart);
+			marker.setAttribute(IMarker.CHAR_END, configMarker.charEnd);
+		} catch (CoreException core) {
+			Platform.getLog(getClass()).log(core.getStatus());
+		}
+	}
+
+	private void removeMarkerFromClangdConfig(IFile configFile) {
+		try {
+			configFile.deleteMarkers(CLANGD_MARKER, false, IResource.DEPTH_INFINITE);
+		} catch (CoreException e) {
+			Platform.getLog(getClass()).log(e.getStatus());
+		}
+	}
+
+	private class ClangdConfigMarker {
+		public String message;
+		public int line = 1;
+		public int charStart = -1;
+		public int charEnd = -1;
+	}
+
+	/**
+	 * Fetch line and char position information from exception to create a marker for the .clangd file.
+	 * @param exception
+	 * @param file
+	 * @return
+	 */
+	private ClangdConfigMarker parseYamlException(MarkedYAMLException exception) {
+		var marker = new ClangdConfigMarker();
+		marker.message = exception.getProblem();
+		marker.line = exception.getProblemMark().getLine() + 1; //getLine() is zero based, IMarker wants 1-based
+		int index = exception.getProblemMark().getIndex();
+		var buffer = exception.getProblemMark().getBuffer();
+		if (index == buffer.length) {
+			index = getIndexOfLastPrintableChar(buffer);
+		}
+		marker.charStart = index;
```

Comment:
This seems hard to solve. As you mentioned above, all we have is the number of bytes:
> That code snippet is the number of bytes in the file :-( number of bytes, chars and code points can all be different.

 > Turning the code point index back into char pr byte index is a non-trivial operation. I would say it probably isn't worth it.

Reply:
Yes - it is a small (but interesting) problem. I am not convinced you need to solve it, but just be aware of it.

You can choose to not solve it, and leave it as a known problem that it will get the charStart/End correct only if  bytes == chars == code points. I don't know how often this will be a real world problem.

Same thing for the above end of file. 

I think it is ok to have this deficiency because the error marker will be present on the file and error message will be correct.

Reply:
PS If you do want to solve the problem, you can read the number of code points in the file and then you will have exact indexes. To do this you need to do BOM handling (see `org.yaml.snakeyaml.reader.UnicodeReader`) and convert char to code point (see `org.yaml.snakeyaml.reader.StreamReader.update()` or `org.yaml.snakeyaml.error.Mark.toCodePoints(char[])` for code you may be able to call)

Reply:
> Yes - it is a small (but interesting) problem. I am not convinced you need to solve it, but just be aware of it.
You can choose to not solve it, and leave it as a known problem that it will get the charStart/End correct only if bytes == chars == code points. I don't know how often this will be a real world problem.

I'll see if I can solve this later. I can open a separate issue for that.

> Same thing for the above end of file.

It is indeed a problem on larger files. I've to re-read the file into a buffer stream to get the end of the buffer. The buffer in the MarkedYAMLException was limited to ~800 bytes:
![image](https://github.com/eclipse-cdt/cdt-lsp/assets/123444711/9bd51007-1f30-4bf0-bd51-2f42dbde6533)


> I think it is ok to have this deficiency because the error marker will be present on the file and error message will be correct.

IMO the missing error marker at the end of the file is more relevant, because I think it is more a real world problem to forget the final } than using non UTF-8 characters in a txt file. So I would like to add the re-read of the file to the catch clause at mentioned [here](https://github.com/eclipse-cdt/cdt-lsp/pull/249#discussion_r1482070539).

Reply:
Sounds good.